src/app/ClusterInfo.h

```diff
@@ -25,96 +25,37 @@
+        VerifyOrReturnError(!mClusterId.HasValue() || mClusterId == other.mClusterId, false);
+        VerifyOrReturnError(!mFieldId.HasValue() || mFieldId == other.mFieldId, false);
+        VerifyOrReturnError(!mListIndex.HasValue() || mListIndex == other.mListIndex, false);
 
-        // If list index exists, field index must exist
-        // Field 0xFFFFFFF (any) &  listindex set is invalid
-        assert(!(myListIndex.HasValue() && !myFieldId.HasValue()));
-        assert(!(otherListIndex.HasValue() && !otherFieldId.HasValue()));
-        assert(!(myFieldId == Optional<AttributeId>::Value(kRootAttributeId) && myListIndex.HasValue()));
-        assert(!(otherFieldId == Optional<AttributeId>::Value(kRootAttributeId) && otherListIndex.HasValue()));
-
-        if (myFieldId == Optional<AttributeId>::Value(kRootAttributeId))
-        {
-            return true;
-        }
-
-        if (myFieldId != otherFieldId)
-        {
-            return false;
-        }
-
-        // We only support top layer for attribute representation, either FieldId or FieldId + ListIndex
-        // Combination: if myFieldId == otherFieldId, ListIndex cannot exist without FieldId
-        // 1. myListIndex and otherListIndex both missing or both exactly the same, then current is superset of other
-        // 2. myListIndex is missing, no matter if otherListIndex is missing or not, then current is superset of other
-        if (myListIndex == otherListIndex)
-        {
-            // either both missing or both exactly the same
-            return true;
-        }
-
-        if (!myListIndex.HasValue())
-        {
-            // difference is ok only if myListIndex is missing
-            return true;
-        }
-
-        return false;
+        return true;
     }
 
+    bool HasWildcard() const { return !mEndpointId.HasValue() || !mClusterId.HasValue() || !mFieldId.HasValue(); }
+
     ClusterInfo() {}
-    NodeId mNodeId         = 0;
-    ClusterId mClusterId   = 0;
-    ListIndex mListIndex   = 0;
-    AttributeId mFieldId   = 0;
-    EndpointId mEndpointId = 0;
-    BitFlags<Flags> mFlags;
-    ClusterInfo * mpNext = nullptr;
-    EventId mEventId     = 0;
-    /* For better structure alignment
-     * Above ordering is by bit-size to ensure least amount of memory alignment padding.
-     * Changing order to something more natural (e.g. clusterid before nodeid) will result
-     * in extra memory alignment padding.
-     * uint64 mNodeId
-     * uint16_t mClusterId
-     * uint16_t mListIndex
-     * uint8_t FieldId
-     * uint8_t EndpointId
-     * uint8_t mDirty
-     * uint8_t mType
-     * uint32_t mpNext
-     * uint16_t EventId
-     * padding 2 bytes
-     */
+
+    Optional<NodeId> mNodeId;
+    Optional<EndpointId> mEndpointId;
+    Optional<ClusterId> mClusterId;
+    Optional<AttributeId> mFieldId;
+    Optional<EventId> mEventId;
+    Optional<ListIndex> mListIndex;
```

Comment:
So... I can see that @mrjerryjohns asked you to do this with Optional, but these are long-lived objects, so we actually care about their size.  This representation is larger than the one that used bitflags, which was in turn larger than the one that uses invalid values.  The .bss numbers for this PR show the increase from this, and that's without us having any subscriptions set up, right?

I feel pretty strongly that we should not be paying the memory hit here, for these non-transient objects.

Reply:
Oh, memory usage, you are right. I did not take the memory usage into consideration before, thanks.

This optional implementation uses 56bytes, while a aligned struct uses 32bytes, that cannot be ignored. @mrjerryjohns

Reply:
Since `ClusterInfo` is fairly internal to the IM, this sounds fine to switch to enum values.